### ISSUE TITLE
Compression

### ISSUE THREAD

**Issue by Msiavashi** [Msiavashi/inference-offloading-doc#3]
# Tiny-Metadata Head-Wise KV Compression (Reproducible Spec)

This doc explains exactly how to compress and decompress a transformer **KV cache** (e.g., layer 15) using a **deterministic, iteration-stable** int8 scheme with **tiny metadata** (per-head base exponent + 2-bit group deltas). Anyone can implement it from scratch.

---

## 1) Scope and notation

* Input tensor per cache (K or V): `X ∈ fp16` with shape **\[B, H, S, D]**.
  Typical example: `B=1`, `H=32`, `S=630`, `D=96`.
* We compress per batch independently; below we assume `B=1` for clarity and work on `X[0] → Xh` with shape **\[H, S, D]**.
* We **group the D dimension** with group size **g** (default `g=16`). Then `G = D / g` groups per head.

We will produce:

* **Codes**: `Q ∈ int8` with shape **\[H, S, D]**
* **Metadata**: tiny, storing the per-head base exponent and 2-bit deltas per group
* **Hyperparameters**: `g`, rounding mode, headroom factor (defaults below)

---

## 2) Rationale and guarantees

* Quantization is **symmetric, mid-rise** around 0 with step `scale = 2^e` per (head, group).
* We store the **exponents** (integers), not floating scales → **tiny metadata** and exact **idempotence** across re-compression.
* After the **first** compress→decompress, repeating the operation with the **same exponents** produces **bit-identical codes** (no drift).

---

## 3) Encoder (compression)

**Inputs**: `X ∈ fp16 [1,H,S,D]`, group size `g`, headroom `α ≥ 1` (default `α = 1.01`)
**Outputs**: `Q ∈ int8 [H,S,D]`, **metadata** = `{base_exp[H], deltas[H,G], header}`

### Step 3.1: Precompute max magnitude per (head, group)

1. Reshape to 4-D groups: `X4 = X.view(1, H, S, G, g)`.

2. For each (h, gidx), compute:

   ```
   amax[h, gidx] = max_{s∈[0..S-1], d∈[0..g-1]} |X4[0,h,s,gidx,d]|
   ```

3. Compute a safety **headroom** scale:

   ```
   raw_scale[h,gidx] = (amax[h,gidx] / 127) * α
   ```

   (α≈1.01 keeps saturation rare.)

4. Convert to **integer exponent** (power of two), using ceil to ensure the step is large enough:

   ```
   e_full[h,gidx] = ceil(log2(max(raw_scale[h,gidx], ε)))
   ```

   with ε = 1e-12.

This gives integer exponents `e_full ∈ ℤ^{H×G}`.

### Step 3.2: Base + 2-bit deltas (tiny metadata)

For each head `h`:

1. Choose a **base exponent**:

   ```
   base_exp[h] = median(e_full[h, :])
   ```

   (Median minimizes total deviation; you can also use mean/round.)

2. Compute **deltas** and clamp to the 2-bit range:

   ```
   delta[h, gidx] = clamp(e_full[h,gidx] - base_exp[h], min=-2, max=+1)
   ```

   Store each `delta` in 2 bits using mapping, e.g.:

   ```
   -2 → 0b00,  -1 → 0b01,  0 → 0b10,  +1 → 0b11
   ```

The **final exponent** used for quantization is:

```
e[h,gidx] = base_exp[h] + delta[h,gidx]
scale[h,gidx] = 2^( e[h,gidx] )
```

> Overflow note: if some `(e_full - base_exp)` are outside \[-2, +1], they’ll be **clamped**. That slightly loosens the local step but keeps metadata tiny. You can track the count for logging.

### Step 3.3: Quantize to int8 (codes)

Using `X4 = X.view(1,H,S,G,g)` and `scale[h,gidx]`:

```
Q4 = clip( round( X4 / scale[None, :, None, :, None] ), -127, 127 )  ∈ int8
Q  = Q4.view(H, S, D)
```

**Store**:

* `Q` (the int8 codes)
* `base_exp` (one signed byte per head if in \[-128,127], otherwise two bytes)
* `delta` packed at **2 bits** per (head, group)
* A tiny header (see §6)

---

## 4) Decoder (decompression)

**Inputs**: `Q ∈ int8 [H,S,D]`, metadata `{base_exp[H], deltas[H,G], g}`, and the original `H, S, D`.

1. Reconstruct exponents per group:

   ```
   e[h,gidx] = base_exp[h] + decode_delta_2bits(delta[h,gidx])
   scale[h,gidx] = 2^( e[h,gidx] )
   ```
2. Reshape `Q4 = Q.view(H,S,G,g)` and dequantize:

   ```
   X̂ = (Q4 * scale[:, None, :, None]).view(H, S, D)  ∈ float32
   ```
3. If you need the original shape `[1,H,S,D]`, prepend batch dimension:

   ```
   X̂_full = X̂.unsqueeze(0)   # [1,H,S,D]
   ```

**Idempotence proof**: `Q` came from rounding `X/scale`. Therefore `(Q*scale)/scale = Q`, and re-rounding yields `Q` again.

---

## 5) File format (recommended)

A compact, explicit format that’s future-proof and endianness-safe.

**Header (little-endian)**

```
bytes 0..3   : magic = b'KVQ1'
bytes 4..5   : version = 0x0001
bytes 6      : group_size (uint8)
bytes 7      : delta_encoding = 0 (00→-2, 01→-1, 10→0, 11→+1)
bytes 8..11  : H (uint32)
bytes 12..15 : S (uint32)
bytes 16..19 : D (uint32)
bytes 20..23 : reserved (zeros)
```

**Metadata**

* `base_exp`: H entries, each **int8** if all in \[-128,127], else **int16** (store a 1-byte flag before them: 0=int8, 1=int16).
* `delta` packed stream: `H*G*2` bits, packed into bytes (row-major over `h`, then `gidx`).
* (Optional) a small “overflow count” (uint32) for diagnostics only.

**Codes**

* `Q` as raw **int8** in row-major `[H,S,D]`. (You can also store `[S,H,D]` if you prefer; just document it.)

You may store **K** and **V** separately or concatenate them (write two sections with magic headers).

---

## 6) Default hyperparameters

* **group\_size g**: 16 (works well for `D=96` → G=6)
* **headroom α**: 1.01
* **rounding**: nearest (ties to even or away from zero—choose and keep consistent; PyTorch `.round()` is banker’s rounding)
* **clip**: saturate to `[-127, +127]`

---

## 7) Complexity and size

* **Time**: O(H·S·D). Max operations are simple divides, rounds, and multiplies.
* **Codes**: exactly **8 bits/elem** (int8).
* **Metadata**:

  * `base_exp`: `H` bytes (usually)
  * `delta`: `H·G·2` bits (2 bits per group)
  * **Header**: \~24–32 bytes
  * For `H=32, D=96, g=16` → `G=6` → `32·6·2 = 384` bits = **48 bytes** for deltas → total per cache ≈ **\~99 bytes** (as measured).
* **Total size**: \~50% of fp16, with \~**0.0002 MB** metadata per cache in the example.

---

## 8) Reference PyTorch code (drop-in)

```python
import torch

def encode_kv(X_fp16, group_size=16, headroom=1.01):
    # X_fp16: [1,H,S,D] float16
    X = X_fp16.float()
    _, H, S, D = X.shape
    assert D % group_size == 0
    G = D // group_size

    # Step 1: per (h,gidx) exponent from amax
    X4 = X.view(1, H, S, G, group_size)
    amax = X4.abs().amax(dim=(0, 2, 4))  # [H,G]
    scale_raw = (amax / 127.0) * headroom
    e_full = torch.ceil(torch.log2(torch.clamp(scale_raw, min=1e-12))).to(torch.int32)

    # Step 2: base + 2-bit delta
    base = torch.median(e_full, dim=1).values.to(torch.int32)   # [H]
    delta = (e_full - base[:, None]).clamp_(-2, 1).to(torch.int8)  # [H,G]
    e = base[:, None].to(torch.int32) + delta.to(torch.int32)   # [H,G]
    scale = torch.pow(2.0, e.float())                           # [H,G]

    # Step 3: quantize to int8
    Q4 = torch.clamp(torch.round(X4 / scale[None, :, None, :, None]), -127, 127).to(torch.int8)
    Q = Q4.view(H, S, D).contiguous()

    # Pack metadata
    base_dtype = torch.int8 if (base.abs() <= 127).all() else torch.int16
    base_stored = base.to(base_dtype)
    # Pack 2-bit deltas into bytes (row-major h,gidx)
    bits = []
    for h in range(H):
        for gidx in range(G):
            d = int(delta[h, gidx].item())
            code = { -2: 0b00, -1: 0b01, 0: 0b10, 1: 0b11 }[d]
            bits.append(code)
    # pack 4 symbols per byte (2 bits each)
    import math
    n = len(bits)
    out = bytearray(math.ceil(n/4))
    for i, sym in enumerate(bits):
        byte_i = i // 4
        off = (i % 4) * 2
        out[byte_i] |= (sym & 0x3) << off
    deltas_packed = bytes(out)

    meta = {
        "H": int(H), "S": int(S), "D": int(D),
        "group_size": int(group_size),
        "headroom_x100": int(round(headroom * 100)),
        "base_exp_dtype": "int8" if base_dtype is torch.int8 else "int16",
        "base_exp": base_stored.cpu().numpy().tobytes(),
        "deltas_2bit": deltas_packed,
    }
    return Q, meta

def decode_kv(Q_int8, meta):
    H = meta["H"]; S = meta["S"]; D = meta["D"]
    g = meta["group_size"]; G = D // g

    # Unpack base
    import numpy as np
    if meta["base_exp_dtype"] == "int8":
        base = torch.frombuffer(memoryview(meta["base_exp"]), dtype=torch.int8).to(torch.int32)
    else:
        base = torch.frombuffer(memoryview(meta["base_exp"]), dtype=torch.int16).to(torch.int32)
    base = base.view(H)

    # Unpack 2-bit deltas
    packed = meta["deltas_2bit"]
    total = H * G
    delta = torch.empty((H, G), dtype=torch.int8)
    mapping = [ -2, -1, 0, +1 ]
    for i in range(total):
        byte_i = i // 4
        off = (i % 4) * 2
        sym = (packed[byte_i] >> off) & 0x3
        delta.view(-1)[i] = mapping[sym]
    e = base[:, None].to(torch.int32) + delta.to(torch.int32)
    scale = torch.pow(2.0, e.float())   # [H,G]

    Q = Q_int8.view(H, S, D)
    Q4 = Q.view(H, S, G, g).float()
    Xhat = (Q4 * scale[:, None, :, None]).view(H, S, D)  # float32
    return Xhat.unsqueeze(0)
```

---

## 9) Evaluation checklist

* **Compression** (relative to fp16):

  * Codes = 8 bpe, metadata ≪ 0.01 bpe → \~**50%** of fp16.
* **Clipping**: fraction of `|Q|==127`. Keep **< 0.5%**. Increase `α` slightly if needed.
* **Iteration stability**:

  * Verify `Q_cycle2 == Q_cycle1` and `Q_cycle3 == Q_cycle1` (byte-for-byte).
* **Error metrics**:

  * MSE / MAE vs original; overall cosine; **per-head cosine** (especially for K).
* **(Optional) K row-norm preservation**:

  * Store per-row norms (e.g., 12-bit log quant) to preserve cosine angles even tighter.

---

## 10) Extensions (optional)

* **int7**: Store 7-bit codes (pack tightly). Combine with tiny metadata → \~**43–45%** of fp16.
* **Residual low-rank patch**: Add a very small per-head rank-2/4 SVD on residuals to recover accuracy at same or lower size.
* **Entropy coding**: Range code the int8 stream; often reduces payload to **\~6.5–7.5 bits/elem** losslessly (distribution-dependent).

---

## 11) Practical tips

* Compute in **fp32**, store/ship codes in **int8** and metadata as bytes.
* Keep rounding mode and mapping consistent across encoder/decoder.
* Persist `g`, `H,S,D`, and the delta mapping in the header for safety.
* If you ever change α, version-bump the header.

---

> https://chatgpt.com/c/6899c082-c618-832f-933a-bdba03da64a8

**Comment by Msiavashi** [comment]
# HF Transformers Experiment

> `$ time CUDA_VISIBLE_DEVICES=1 lm_eval --model hf   --model_args pretrained=microsoft/Phi-3-mini-4k-instruct,add_bos_token=true,dtype=float16   --tasks gsm8k   --num_fewshot 5   --limit 100   --batch_size "auto" --seed 42`

> group_size = 16

## Default HF

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.83|±  |0.0378|
|     |       |strict-match    |     5|exact_match|↑  | 0.83|±  |0.0378|


real    2m21.944s
user    2m1.361s
sys     0m22.564s

[gpu_mem_log_default.csv](https://github.com/user-attachments/files/21776842/gpu_mem_log_default.csv)

## Compress and Decompress at every layer

We naively compress and decompress immediately:
```
        comp_k, comp_v, k_meta, v_meta = compress(key_states, value_states)
        key_states, value_states = decompress(comp_k, comp_v, k_meta, v_meta)
```

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.72|±  |0.0451|
|     |       |strict-match    |     5|exact_match|↑  | 0.70|±  |0.0461|


real    3m58.925s
user    3m16.969s
sys     0m42.677s

[gpu_mem_log.csv](https://github.com/user-attachments/files/21776838/gpu_mem_log.csv)

**Comment by Msiavashi** [comment]
## Compress and Decompress

> group_size = 96   # as large as D (heads dimention)

> `$ time CUDA_VISIBLE_DEVICES=1 lm_eval --model hf   --model_args pretrained=microsoft/Phi-3-mini-4k-instruct,add_bos_token=true,dtype=float16   --tasks gsm8k   --num_fewshot 5   --limit 100   --batch_size "auto" --seed 42`

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.84|±  |0.0368|
|     |       |strict-match    |     5|exact_match|↑  | 0.82|±  |0.0386|


real    3m54.559s
user    3m10.870s
sys     0m43.428s

[gpu_mem_log.csv](https://github.com/user-attachments/files/21777512/gpu_mem_log.csv)

## Default HF


|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  | 0.83|±  |0.0378|
|     |       |strict-match    |     5|exact_match|↑  | 0.83|±  |0.0378|


real    2m25.507s
user    2m0.829s
sys     0m23.114s

[gpu_mem_log_default.csv](https://github.com/user-attachments/files/21777510/gpu_mem_log_default.csv)

*Conclusion*: Increasing the group size leads to less metadata and slightly faster processing. Surprisingly, in this case, it significantly improves accuracy. However, reducing the group size to 8 results in a drop in accuracy to approximately 62%. The reasons for this need to be explored further.


<img width="1694" height="1140" alt="Image" src="https://github.com/user-attachments/assets/297b2c11-6739-443d-a0e9-69ca37baa13c" />
